frontend/src/metabase/modes/types.ts

```diff
@@ -0,0 +1,102 @@
+import React from "react";
+import type {
+  DatasetColumn,
+  RowValue,
+  Series,
+  VisualizationSettings,
+} from "metabase-types/api";
+import type { Dispatch, ReduxAction } from "metabase-types/store";
+import type { OnChangeCardAndRun } from "metabase/visualizations/types";
+import type Question from "metabase-lib/Question";
+
+type DimensionValue = {
+  value: RowValue;
+  column: DatasetColumn;
+};
+
+export type ClickObject = {
+  value?: RowValue;
+  column?: DatasetColumn;
+  dimensions?: DimensionValue[];
+  settings?: Record<string, unknown>;
+  extraData?: Record<string, unknown>;
+  seriesIndex?: number;
+  origin?: {
+    row: RowValue;
+    cols: DatasetColumn[];
+  };
+  event?: MouseEvent;
+  element?: HTMLElement;
+};
+
+type Dispatcher = (dispatch: Dispatch) => void;
+
+export type ClickActionPopoverProps = {
+  series: Series;
+  onChangeCardAndRun: OnChangeCardAndRun;
+  onChange: (settings: VisualizationSettings) => void;
+  onResize: (...args: unknown[]) => void;
+  onClose: () => void;
+};
+
+export type ClickActionButtonType =
+  | "formatting"
+  | "horizontal"
+  | "info"
+  | "token"
+  | "token-filter"
+  | "sort";
+
+export type ClickActionBase = {
+  name: string;
+  title?: React.ReactNode;
+  section: string;
+  icon?: string;
+  buttonType: ClickActionButtonType;
+  default?: boolean;
+  defaultAlways?: boolean;
+  tooltip?: string;
+  extra?: () => Record<string, unknown>;
+};
+
+type ReduxClickAction = {
+  action: () => ReduxAction | Dispatcher;
+};
+
+type QuestionChangeClickAction = {
+  question: () => Question;
+};
+
+type PopoverClickAction = {
+  popoverProps?: Record<string, unknown>;
+  popover: (props: ClickActionPopoverProps) => JSX.Element;
+};
+
+type UrlClickAction = {
+  ignoreSiteUrl?: boolean;
+  url: () => string;
+};
+
+export type RegularClickActionEffect =
+  | ReduxClickAction
+  | QuestionChangeClickAction
+  | PopoverClickAction
+  | UrlClickAction;
+
+type RegularClickAction = ClickActionBase & RegularClickActionEffect;
```

Comment:
Could we remove this type if we extend specific action types?

Reply:
We're using it down below for the [`AlwaysDefaultClickAction` type](https://github.com/metabase/metabase/blob/a61ed7d2ae4053e5e00ac1015dded683b4f3cf6a/frontend/src/metabase/modes/types.ts#L85).

Any action with `defaultAlways: true` is treated as one. Having an action like that, Metabase will ignore all the other possible drills and immediately execute the always default one. AFAIK it was introduced for the click behavior drill, which should act exactly that way. TBH it feels this behavior "overwrite" should happen on the dashboard level, but fixing this feels out of the scope of this change.

Any click action (URL / redux action / etc.) can be an "always default". Important point here is that for the "always default" action, we can skip the UI-related properties like section, icon, title, etc. as there's no popup to show them in. So we extend the `RegularClickAction` type [here](https://github.com/metabase/metabase/blob/a61ed7d2ae4053e5e00ac1015dded683b4f3cf6a/frontend/src/metabase/modes/types.ts#L86) and omit all the irrelevant properties. And then the final [`ClickAction` type](https://github.com/metabase/metabase/blob/a61ed7d2ae4053e5e00ac1015dded683b4f3cf6a/frontend/src/metabase/modes/types.ts#L92) is either a regular action that can be presented in a popup OR an "always default" action

I've tried to simplify this a bit and address other feedback, would appreciate if you can take another look at this @ranquild